src/main/drivers/dshot.c

```diff
@@ -185,117 +184,96 @@ static uint32_t dshotDecodeErpmTelemetryValue(uint16_t value)
+        // motor won't fit into debug array
+        return;
     }
+
+    uint16_t highByte = 0;
+    switch (debugType) {
+    case DEBUG_DSHOT_RPM_TELEMETRY:  // eprm value is 16bit, all of debug is used
+    case DEBUG_DSHOT_STATUS:         // store status as value, no need to duplicate it in upper byte (TODO: store upper 4 status bits)
+        break;
+    default:
+        // piggyback status byte on 8-bit telemetry values
+        highByte = dshotTelemetryState.motorState[motorIndex].telemetryData[DSHOT_TELEMETRY_TYPE_STATUS] << (sizeof(uint8_t) * 8);
+    }
+    DEBUG_SET(debugType, motorIndex, highByte | decodedValue);
 }
 
-static bool dshotDecodeTelemetryValue(uint8_t motorIndex, uint32_t *pDecoded, dshotTelemetryType_t *pType)
+// for reference, see https://github.com/bird-sanctuary/extended-dshot-telemetry/blob/main/README.md
+// decode value and type, retrun true if successful
+static bool dshotDecodeTelemetryValue(unsigned motorIndex, uint32_t *pDecoded, dshotTelemetryType_e *pType)
 {
-    bool useDebug = false;
+    const uint16_t rawValue = dshotTelemetryState.motorState[motorIndex].rawValue;
+    uint32_t decoded;
+    dshotTelemetryType_e type;
     debugType_e debugType;
-    const uint16_t value = dshotTelemetryState.motorState[motorIndex].rawValue;
 
     *pDecoded = DSHOT_TELEMETRY_INVALID;
-    if (dshotTelemetryState.motorState[motorIndex].telemetryTypes == DSHOT_NORMAL_TELEMETRY_MASK) {   /* Check DSHOT_TELEMETRY_TYPE_eRPM mask */
-        // Decode eRPM telemetry
-        *pDecoded = dshotDecodeErpmTelemetryValue(value);
-
-        // Update debug buffer (motorIndex < motorCount guaranteed by caller)
-        if (motorIndex < DEBUG16_VALUE_COUNT) {
-            DEBUG_SET(DEBUG_DSHOT_RPM_TELEMETRY, motorIndex, *pDecoded);
+    if (!dshotTelemetryState.motorState[motorIndex].extendedTelemetryEnabled // no extended telemetry
+        || (rawValue & 0x100)               // normalized mantissa (topmost mantissa bit is set)
+        || (rawValue & 0xe00) == 0          // zero exponent
+        ) {
+        // basic or extended telemetry eRPM packet
+        decoded = dshotDecodeErpmTelemetryValue(rawValue);
+        type = DSHOT_TELEMETRY_TYPE_eRPM;
+        dshotStoreDebug(motorIndex, decoded, DEBUG_DSHOT_RPM_TELEMETRY);
+        if (dshotTelemetryState.motorState[motorIndex].extendedTelemetryEnabled) {
+            // allow tracing (reduced precission) RPM + status
+            dshotStoreDebug(motorIndex, (decoded / 18) & 0xff, DEBUG_DSHOT_STATUS_ERPM18);
         }
-
-        *pType = DSHOT_TELEMETRY_TYPE_eRPM;
     } else {
-        // Decode Extended DSHOT telemetry
-        *pDecoded = value & DSHOT_TELEMETRY_VALUE_MASK;
-
-        switch (value & DSHOT_TELEMETRY_RANGE_MASK) {
-
-        case DSHOT_TELEMETRY_RANGE_TEMPERATURE:
+        // Extended DSHOT telemetry packet
+        // map range to corresponding types
+        // TODO: this mapping is 1:1, we can easily use index directly
```

Comment:
Just to clarify: Instead of shifting the range you suggest using `telemetry` as index directly? Getting rid of the `range` parameter and `dshotTelemetryRange_e` enum altogether?

Reply:
as long as dshotTelemetryType_e is limited to 8 values, DSHOT_TELEMETRY_RANGE_SHIFT can be used to convert it to and from range.

if telemetry is ever extended, we can handle 8 edr values as special case.